crates/wysiwyg/src/dom/nodes/text_node.rs

```diff
@@ -96,6 +106,68 @@ where
             false
         }
     }
+
+    /// This gets the character at the cursor, considering the
+    /// direction of travel
+    pub fn char_at_offset(
+        &self,
+        offset: usize,
+        direction: &Direction,
+    ) -> Option<char> {
+        self.data()
+            .chars()
+            .nth(direction.get_index_from_cursor(offset))
+    }
+
+    /// This gets the character type at the cursor, considering the
+    /// direction of travel
+    pub fn char_type_at_offset(
+        &self,
+        offset: usize,
+        direction: &Direction,
+    ) -> Option<CharType> {
+        let char = self.char_at_offset(offset, direction);
+        match char {
+            Some(c) => Some(get_char_type(c)),
+            None => None,
+        }
+    }
+
+    /// When moving through a node, the cursor counts as inside the node
+    /// at one end, but not the other. This function determines that.
+    pub fn offset_is_inside_node(
+        &self,
+        current_offset: usize,
+        direction: &Direction,
+    ) -> bool {
+        let node_length = self.data().len();
+        match direction {
+            Direction::Forwards => current_offset < node_length,
+            Direction::Backwards => current_offset > 0,
+        }
+    }
+
+    /// Required due to zero length text node existence
+    pub fn has_length(&self) -> bool {
```

Comment:
It's more normal in Rust to use the opposite sense and call it `is_empty`.

After this change is merged, please can you check whether some of it can be simplified since I recently committed changes that reduce the number of empty text nodes you will see in the wild (possibly to zero..?)

Reply:
Yep sure - I will do that as part of the main pr for functionality